tezos/new_context/src/gc/new_gc.rs

```diff
@@ -0,0 +1,289 @@
+use std::{cell::{Ref, RefCell}, collections::{BTreeMap, HashMap}, sync::Arc};
+
+use crossbeam_channel::{Receiver, Sender};
+use serde::{Serialize, Deserialize};
+
+use crate::{hash::EntryHash, working_tree::Entry};
+
+use tezos_spsc::{Consumer, Producer};
+
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct HashId(usize);
+
+#[derive(Debug)]
+pub struct HashInterner {
+    hashes: slab::Slab<EntryHash>,
+    // values: Vec<Option<Arc<[u8]>>>,
+    frees: Consumer<HashId>,
+    values: MapValues,
+}
+
+// pub(crate) struct HashInterner {
+//     inner: RefCell<HashInternerInner>
+// }
+
+impl HashInterner {
+    fn new(consumer: Consumer<HashId>) -> Self {
+        Self {
+            // inner: RefCell::new(HashInternerInner {
+                hashes: slab::Slab::default(),
+                frees: consumer,
+                values: MapValues::default(),
+            // })
+        }
+    }
+
+    pub(crate) fn put(&mut self, hash: EntryHash) -> HashId {
+        // let mut inner = self.inner.borrow_mut();
+        let free_id = self.frees.pop().ok();
+
+        let res = if let Some(free_id) = free_id {
+            self.values.remove(free_id);
+            self.hashes[free_id.0] = hash;
+            free_id
+        } else {
+            let id = self.hashes.insert(hash);
+            HashId(id)
+        };
+
+        res
+    }
+
+    pub(crate) fn get(&self, hash_id: HashId) -> Option<&EntryHash> {
+        self.hashes.get(hash_id.0)
+    }
+
+    pub(crate) fn get_value(&self, hash_id: HashId) -> Option<&[u8]> {
+        self.values.get(hash_id)
+    }
+}
+
+/// Kind of a secondary map HashId -> Arc<[u8]>
+#[derive(Debug, Default)]
+struct MapValues {
+    values: Vec<Option<Arc<[u8]>>>,
+}
+
+impl MapValues {
+    fn set(&mut self, hash_id: HashId, value: Arc<[u8]>) {
+        let index = hash_id.0 as usize;
+
+        if index >= self.values.len() {
+            self.values.resize_with(index + 1, Default::default);
+        }
+
+        self.values[index] = Some(value);
+    }
+
+    fn get(&self, hash_id: HashId) -> Option<&[u8]> {
+        let index = hash_id.0 as usize;
+
+        self.values.get(index)?.as_ref().map(AsRef::as_ref)
+    }
+
+    fn remove(&mut self, free_id: HashId) {
+        let index = free_id.0;
+        self.values[index] = None;
+    }
+}
+
+pub struct NewGC {
+    current: BTreeMap<HashId, Arc<[u8]>>,
```

Comment:
What is the reason for the values being `Arc`? No two hashes will map to the same value.

Reply:
Another question, can we switch later to using the two in-memory backends we already here instead of `BTreeMap` directly?

Reply:
Oh wait, the `Arc` here is the hash, and not the value right?

Reply:
@tizoc The `Arc<[u8]>` here is the serialized value, not the hash.
It uses `Arc` instead of `Box` so that the value is readable from both the GC thread and the main thread.
So both threads can deserialize the value.

When the main thread wants to read a serialized value, it reads it from `MapValues` https://github.com/sebastiencs/tezedge/blob/deduplicate-hashes/tezos/new_context/src/gc/new_gc.rs#L63.

`current: BTreeMap<HashId, Arc<[u8]>>` contains the `HashId -> value` used in the current cycle.
`MapValues` contains `HashId -> value` of all cycles.

> Another question, can we switch later to using the two in-memory backends we already here instead of BTreeMap directly?

Yes, though the 2 memory backends (BTreeMap & HashMap) will replace the `MapValues`, though it will requires some changes